pkg/controller/deployment/deployment_controller.go

```diff
@@ -816,30 +834,131 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
-		glog.V(4).Infof("Pod expectations not met yet before reconciling old RCs\n")
+
+	// Clean up unhealthy replicas first, otherwise unhealthy replicas will block deployment
+	// and cause timeout. See https://github.com/kubernetes/kubernetes/issues/16737
+	cleanupCount, err := dc.cleanupUnhealthyReplicas(oldRCs, deployment, maxScaledDown)
+	if err != nil {
 		return false, nil
 	}
+
+	totalScaledDown := cleanupCount
+	if cleanupCount < maxScaledDown {
+		// Scale down old rcs, need check maxUnavailable to ensure we can scale down
+		scaledDownCount, err := dc.scaleDownOldRCsForRollingUpdate(allRCs, oldRCs, deployment, maxScaledDown-cleanupCount)
+		if err != nil {
+			return false, nil
+		}
+		totalScaledDown = cleanupCount + scaledDownCount
+	}
+
+	if expectationsCheck {
+		dc.podExpectations.ExpectDeletions(dKey, totalScaledDown)
+	}
+
+	return totalScaledDown > 0, nil
+}
+
+// cleanupUnhealthyReplicas will scale down old rcs with unhealthy replicas, so that all unhealthy replicas will be deleted.
+func (dc *DeploymentController) cleanupUnhealthyReplicas(oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxCleanupCount int) (int, error) {
+	sort.Sort(controller.ControllersByCreationTimestamp(oldRCs))
+	// Safely scale down all old rcs with unhealthy replicas. ReplicationController/ReplicaSet will sort the pods in the order
+	// such that not-ready < ready, unscheduled < scheduled, and pending < running. This ensures that unhealthy replicas will
+	// been deleted first and won't increase unavailability.
+	totalScaledDown := 0
+	for _, targetRC := range oldRCs {
+		if totalScaledDown >= maxCleanupCount {
+			break
+		}
+
+		if targetRC.Spec.Replicas == 0 {
+			// cannot scale down this RC.
+			continue
+		}
+		readyPodCount, err := deploymentutil.GetAvailablePodsForRCs(dc.client, []*api.ReplicationController{targetRC}, 0)
+		if err != nil {
+			return totalScaledDown, fmt.Errorf("could not find available pods: %v", err)
+		}
+		if targetRC.Spec.Replicas == readyPodCount {
+			// no unhealthy replicas found, no scaling required.
+			continue
+		}
+
+		scaledDown := int(math.Min(float64(maxCleanupCount-totalScaledDown), float64(targetRC.Spec.Replicas-readyPodCount)))
+		newReplicasCount := targetRC.Spec.Replicas - scaledDown
+		_, err = dc.scaleRCAndRecordEvent(targetRC, newReplicasCount, deployment)
+		if err != nil {
+			return totalScaledDown, err
+		}
+		totalScaledDown += scaledDown
+	}
+	return totalScaledDown, nil
+}
+
+// scaleDownOldRCsForRollingUpdate scales down old rcs when deployment strategy is "RollingUpdate".
+// Need check maxUnavailable to ensure availability
+func (dc *DeploymentController) scaleDownOldRCsForRollingUpdate(allRCs []*api.ReplicationController, oldRCs []*api.ReplicationController, deployment extensions.Deployment, maxScaleDownCount int) (int, error) {
+	maxUnavailable, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxUnavailable)
```

Comment:
As I commented here: https://github.com/kubernetes/kubernetes/pull/20273/files#r51953975
it would be useful to factor this out into a helper function (in pkg/util/deployment).